source/geod24/concurrency.d

```diff
@@ -1446,3 +1502,665 @@ private:
+        {
+            ChannelQueueEntry q_ent = enqueueEntry(this.readq, &output);
+            this.lock.unlock_nothrow();
+
+            q_ent.sem.wait();
+            return q_ent.success;
+        }
+
+        this.lock.unlock_nothrow();
+        return success;
+    }
+
+    /***********************************************************************
+
+        Try to read a message from Channel without blocking
+
+        tryRead reads a message if it is possible to do so without blocking.
+
+        If the tryRead is being executed in a select call, it tries to consume the
+        `caller_sel` with the given `caller_sel_id`. It only proceeds if it can
+        successfully consume the `caller_sel`
+
+        Param:
+            output = Field to write the message to
+            caller_sel = SelectState instace of the select call being executed
+            caller_sel_id = id of the select call being executed
+
+        Returns:
+            Success/Failure
+
+    ***********************************************************************/
+
+    private bool tryRead (ref T output) nothrow
+    {
+        ChannelQueueEntry write_ent = dequeueEntry(this.writeq);
+
+        if (this.buffer.length > 0)
+        {
+            output = this.buffer.front();
+            this.buffer.popFront();
+
+            if (write_ent)
+            {
+                this.buffer.putBack(*(write_ent.pVal));
+            }
+        }
+        // if dequeueEntry returns a valid entry, it always successfully consumes the related select states.
+        // the race between 2 select calls is resolved in dequeueEntry.
+        else if (write_ent)
+        {
+            output = *(write_ent.pVal);
+        }
+        else
+        {
+            return false;
+        }
+
+        if (write_ent)
+            write_ent.sem.notify();
+        return true;
+    }
+
+    /***********************************************************************
+
+        Close the channel
+
+        Closes the channel by marking it closed and flushing all the wait
+        queues
+
+    ***********************************************************************/
+
+    void close () nothrow
+    {
+        this.lock.lock_nothrow();
+        if (!this.closed)
+        {
+            this.closed = true;
+
+            // Wake blocked Fibers up, report the failure
+            foreach (ref entry; this.readq)
+            {
+                entry.terminate();
+            }
+            foreach (ref entry; this.writeq)
+            {
+                entry.terminate();
+            }
+
+            this.readq.clear();
+            this.writeq.clear();
+        }
+        this.lock.unlock_nothrow();
+    }
+
+    /***********************************************************************
+
+        Return the length of the internal buffer
+
+    ***********************************************************************/
+
+    size_t length () nothrow
+    {
+        this.lock.lock_nothrow();
```

Comment:
@omerfirmak my concurrency-foo isn't the best.

Can you help me understand the purpose of using a lock when looking up a value like this? Because as soon as the function returns the lock is released, which means the state of the channel could change while the calling code still assumes the channel is in the same state.

Isn't it the same to just return the value directly?

Reply:
Also I feel like for checking lengths and booleans we should use `atomicStore` and `atomicLoad`, because locking is expensive.

Reply:
You are right. As long as every operation (load/store) on the variables are atomic, there is no need to use a lock in these getters. We can do this easily with the closed. But since the length is managed by the ring buffer implementation, I did not want to prematurely optimize. We can definitely get rid of those once the implementation is complete, feature and correction wise.

Reply:
Good idea. Safety first.